db/commitlog/commitlog.cc

```diff
@@ -422,6 +460,60 @@ class db::commitlog::segment_manager : public ::enable_shared_from_this<segment_
     seastar::semaphore _reserve_recalculation_guard;
 };
 
+future<> db::commitlog::segment_manager::named_file::open(open_flags flags, file_open_options opt) noexcept {
+    assert(!*this);
```

Comment:
This assert suggests that instead of a default constructor and one open() call you must do later, you can have a constructor which takes the open parameters, or alternatively , a private constructor and an open() factory (static) function - to get compile-time sanity checking instead of this runtime assert crashes.
But maybe I'm missing something in the big picture that led you to this design?

Reply:
Recycling. We keep a bunch of named_files in a list, pick one to re-use. These are closed (the seastar::file part is empty). We rename using stored file path, then re-open the object. Could of course just create new objects, using a static open etc. But re-using the object is the motivator here. Assert just means we must call close before calling open again.